bfver

```diff
@@ -48,30 +49,92 @@ Options:
+  --file,-f FILE         print BSP version information from a BFB file instead
+                         of what is installed.
 EOF
 }
 
-print_bootloader_ver () {
+print_path_vers () {
+    # Print the versions of an image at a given path.
+    # For any file, the versions contained in the BFB will be printed.
+    # For /dev/mmcblk0boot*, mlxbf-bootctl will be employed to read the
+    # partition.
+    STREAM_PATH="$1"
+
+    case $STREAM_PATH in
+        /dev/mmcblk0boot*) print_bfb_installed_vers "$STREAM_PATH";;
+        *)
+            if ! [ -e $STREAM_PATH ]; then
+                echo "$PROGNAME: warn: $STREAM_PATH does not exist, skipping" >&2
+            elif ! [ -f $STREAM_PATH ]; then
+                echo "$PROGNAME: warn: $STREAM_PATH is not file, skipping" >&2
+            else
+                print_bfb_file_vers "$STREAM_PATH"
+            fi
+            ;;
+    esac
+}
+
+print_bfb_installed_vers () {
     DEV_PATH=$1
     CURRENT_BFB="$TMP_DIR/$(basename $DEV_PATH).bfb"
 
     # Retrieve default.bfb from the eMMC
     mlxbf-bootctl -r "$DEV_PATH" -b "$CURRENT_BFB" 2>&1 >/dev/null
 
-    echo "--$DEV_PATH"
+    print_bfb_file_vers "$CURRENT_BFB" "$DEV_PATH"
+}
+
+print_bfb_file_vers () {
+    # Print versions stored in files.
+    BFB_PATH="$1"
+
+    # Second argument is optional, specifies a display name for the file.
+    # Defaults to $1.
+    if [ -n "$2" ]; then
+        DISPLAY_NAME="$2"
+    else
+        DISPLAY_NAME="$BFB_PATH"
+    fi
+
+    echo "--$DISPLAY_NAME"
+
+    # Find and print the ATF version
+    echo BlueField ATF version: "$(strings "$BFB_PATH" | grep -m 1 "(\(release\|debug\))")"
+
+    # Keep UEFI version and use it to determine BSP version as well
+    UEFI_VER="$(strings -el "$BFB_PATH" | grep "BlueField" | cut -d':' -f 2)"
 
-    # Find and print the ATF version of current bfb
-    echo BlueField ATF version: "$(strings "$CURRENT_BFB" | grep -m 1 "(\(release\|debug\))")"
+    echo BlueField UEFI version: $UEFI_VER
 
-    # Find and print the UEFI of current bfb
-    echo BlueField UEFI version: "$(strings -e l "$CURRENT_BFB" | grep "BlueField" | cut -d':' -f 2)"
+    # sed regex needs to be escaped heavily, so in more normal terms:
+    # (.+\..+\..+).*-[0-9]+-g[0-9a-fA-F]+
+    # Essentially, match the string as Major.Minor.Patch in front,
+    # the git describe tag from the back, and anything in the middle.
+    BSP_MAJOR_MINOR_PATCH="$(echo "$UEFI_VER" | sed 's/\(.\+\..\+\..\+\).*-[0-9]\+-g[0-9a-fA-F]\+/\1/')"
+
+    # Find build ID
+    BUILD_ID="$(strings -el "$BFB_PATH" | grep "BId" | sed "s/^BId//")"
+
+    if [ -z "$BUILD_ID" ]; then
+        echo "$PROGNAME: warn: could not find UEFI build ID (likely bootloader too old)" >&2
+        BUILD_ID="NO_BUILD_ID"
+    fi
+
+    if [ "$BUILD_ID" = 0 ]; then
+        echo "$PROGNAME: warn: build id of 0 (likely development image)" >&2
+    fi
+
+    echo BlueField BSP version: ${BSP_MAJOR_MINOR_PATCH}.$BUILD_ID
```

Comment:
Do we want to print $BUILD_ID if BUILD_ID="NO_BUILD_ID"?

I am just concerned whether it would break any regression (obviously not our regression but QA/verification side)